core/src/raw/oio/write/exact_buf_write.rs

```diff
@@ -59,15 +59,8 @@ impl<W: oio::Write> oio::Write for ExactBufWriter<W> {
     }
 
     async fn close(&mut self) -> Result<()> {
-        loop {
-            if self.buffer.is_empty() {
-                break;
-            }
-
-            let written = self.inner.write(self.buffer.collect()).await?;
-            self.buffer.advance(written);
-        }
-
+        let buffer = std::mem::take(&mut self.buffer);
```

Comment:
I'm not sure if we need loop still. Generally `&mut self` should take the ref exclusive. But I'm not quit sure.

Reply:
If `.await` can give up the mut ref, we always have the risk the before `close`, there is a new push.

Reply:
`Close` could fail and users might retry `close` later, we should keep buffer in memory until users decide to drop this writer.